lib/action_cable/subscription_adapter/sqlite.rb

```diff
@@ -0,0 +1,155 @@
+# frozen_string_literal: true
+
+require 'action_cable/subscription_adapter/async'
+
+module ActionCable
+  module SubscriptionAdapter
+    class Sqlite < Async
+      def initialize(*)
+        super
+        @sync = Mutex.new
+        @listener = Listener.new(self)
+      end
+
+      # alias the existing broadcast method from Async adapter
+      # so that we can call it from the listener
+      alias broadcast_to_clients broadcast
+
+      # overwrite the broadcast so that messages are stored
+      # in sqlite, and let the listener pick them up and actually
+      # broadcast them
+      def broadcast(channel, payload)
+        # must synchronize writes to sqlite because it does not
+        # support concurrent writes to the same table
+        @sync.synchronize do
+          with_broadcast_connection do |sqlite_conn|
+            sqlite_conn.execute(
+              'INSERT INTO messages VALUES (?, ?, ?)',
+              nil,
+              channel,
+              JSON.dump(payload)
+            )
+          end
+        end
+      end
+
+      def shutdown
+        @listener.shutdown
+      end
+
+      def with_subscriptions_connection
+        ar_conn = nil
+        ActionCableRecord.connected_to(role: :reading) do
+          ar_conn = ActionCableRecord.connection_pool.checkout.tap do |conn|
+            # Action Cable is taking ownership over this database connection, and
+            # will perform the necessary cleanup tasks
+            ActionCableRecord.connection_pool.remove(conn)
+          end
+          sqlite_conn = ar_conn.raw_connection
+          verify!(sqlite_conn)
+          sqlite_conn.results_as_hash = true
+          yield sqlite_conn
+        end
+      ensure
+        ar_conn&.disconnect!
+      end
+
+      def with_broadcast_connection
+        ActionCableRecord.connected_to(role: :writing) do
+          ActionCableRecord.connection_pool.with_connection do |conn|
+            sqlite_conn = conn.raw_connection
+            verify!(sqlite_conn)
+            yield sqlite_conn
+          end
+        end
+      end
+
+      private
+
+      def listener
+        @listener || @server.mutex.synchronize { @listener ||= Listener.new(self) }
+      end
+
+      def verify!(sqlite_conn)
+        return if sqlite_conn.is_a?(SQLite3::Database)
+
+        raise 'The Active Record database must be SQLite in order to use the SQLite Action Cable storage adapter'
+      end
+
+      class Listener
+        def initialize(adapter)
+          super()
+
+          @adapter = adapter
+
+          @poll_interval = 0.02
+          @shutdown = false
+          @latest_message_id = nil
+
+          @thread = Thread.new do
```

Comment:
Need to wrap this is an [application executor](https://guides.rubyonrails.org/threading_and_code_execution.html#wrapping-application-code)